services/horizon/internal/ingest/contractevents/events.go

```diff
@@ -0,0 +1,376 @@
+	// Format: ["burn", from addr, sep11 asset]
+	if len(topics) != 3 {
+		return errors.New("burn event requires 3 topics")
+	}
+
+	from, err := parseAddress(topics[1])
+	if err != nil {
+		return fmt.Errorf("invalid from address: %w", err)
+	}
+	event.From = from
+	return nil
+}
+
+func parseMintEventFromTxMetaV3(topics xdr.ScVec, event *StellarAssetContractEvent) error {
+	// Format: ["mint", admin addr, to addr, sep11 asset], i128 amount
+	if len(topics) != 4 {
+		return errors.New("mint event requires 4 topics")
+	}
+
+	// Admin is not used. but needs to be parsed for SAC format correctness
+	_, err := parseAddress(topics[1])
+	if err != nil {
+		return fmt.Errorf("invalid admin address: %w", err)
+	}
+	to, err := parseAddress(topics[2])
+	if err != nil {
+		return fmt.Errorf("invalid to address: %w", err)
+	}
+	event.To = to
+	return nil
+}
+
+func parseMintEventFromTxMetaV4(topics xdr.ScVec, event *StellarAssetContractEvent) error {
+	// Format: ["mint", to addr, sep11 asset] - NO admin address in V4
+	if len(topics) != 3 {
+		return errors.New("mint event requires 3 topics")
+	}
+
+	to, err := parseAddress(topics[1])
+	if err != nil {
+		return fmt.Errorf("invalid to address: %w", err)
+	}
+	event.To = to
+	return nil
+}
+
+func parseClawbackEventFromTxMetaV3(topics xdr.ScVec, event *StellarAssetContractEvent) error {
+	// Format: ["clawback", admin addr, from addr, sep11 asset], i128 amount
+	if len(topics) != 4 {
+		return errors.New("clawback event requires 4 topics")
+	}
+
+	// Admin is not used. but needs to be parsed for SAC format correctness
+	_, err := parseAddress(topics[1])
+	if err != nil {
+		return fmt.Errorf("invalid admin address: %w", err)
+	}
+	from, err := parseAddress(topics[2])
+	if err != nil {
+		return fmt.Errorf("invalid from address: %w", err)
+	}
+	event.From = from
+	return nil
+}
+
+func parseClawbackEventFromTxMetaV4(topics xdr.ScVec, event *StellarAssetContractEvent) error {
+	// Format: ["clawback", from addr, sep11 asset] - NO admin address in V4
+	if len(topics) != 3 {
+		return errors.New("clawback event requires 3 topics")
+	}
+
+	from, err := parseAddress(topics[1])
+	if err != nil {
+		return fmt.Errorf("invalid from address: %w", err)
+	}
+	event.From = from
+	return nil
+}
+
+func parseSacEventFromTxMetaV3(event *xdr.ContractEvent, networkPassphrase string) (*StellarAssetContractEvent, error) {
+	eventType, asset, topics, data, err := parseCommonEventValidation(event, networkPassphrase)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse amount (V3 is always direct i128)
+	amount, ok := data.GetI128()
+	if !ok {
+		return nil, fmt.Errorf("invalid amount in event data: %v", data.String())
+	}
+
+	// Parse addresses based on event type
+	sacEvent := &StellarAssetContractEvent{
+		Type:   eventType,
+		Asset:  asset,
+		Amount: amount,
+	}
+
+	switch eventType {
+	case EventTypeTransfer:
+		if err := parseTransferEvent(topics, sacEvent); err != nil {
+			return nil, err
+		}
```

Comment:
each of these could be 
```suggestion
		return sacEvent, parseTransferEvent(topics, sacEvent)
```
since `parse<Type>Event` modifies `sacEvent` by reference, right?

Reply:
Alternatively, just assign `err` in each `case` and only `if err != nil` at the end.

Reply:
it will have to be the latter, since the parseXXXEvent field return type is `error` instead of `(*StellarAssetContractEvent, error) `